src/zcash/Address.cpp

```diff
@@ -36,6 +38,47 @@ SproutPaymentAddress SproutSpendingKey::address() const {
     return viewing_key().address();
 }
 
+//! Sapling
+SaplingFullViewingKey SaplingExpandedSpendingKey::full_viewing_key() const {
+    uint256 ak;
+    uint256 nk;
+    librustzcash_ask_to_ak(ask.begin(), ak.begin());
+    librustzcash_nsk_to_nk(nsk.begin(), nk.begin());
+    return SaplingFullViewingKey(ak, nk, ovk);
+}
+
+SaplingExpandedSpendingKey SaplingSpendingKey::expanded_spending_key() const {
+    return SaplingExpandedSpendingKey(PRF_ask(*this), PRF_nsk(*this), PRF_ovk(*this));
+}
+
+SaplingFullViewingKey SaplingSpendingKey::full_viewing_key() const {
+    return expanded_spending_key().full_viewing_key();
+}
+
+SaplingIncomingViewingKey SaplingFullViewingKey::in_viewing_key() const {
+    uint256 ivk;
+    librustzcash_crh_ivk(ak.begin(), nk.begin(), ivk.begin());
+    return SaplingIncomingViewingKey(ivk);
+}
+
+SaplingSpendingKey SaplingSpendingKey::random() {
+    return SaplingSpendingKey(random_uint256());
```

Comment:
Per section 4.2.2 of the Sapling spec, we need to discard spending keys that result in `ivk == 0`, and loop around. This happens very infrequently however, and we'll end up not using this function if we get ZIP 32 into the wallet, so fixing this is non-blocking (but must happen before 2.0.0 if we don't get ZIP 32 in). Either address here, or open an issue to address later.

Reply:
So return an option from `in_viewing_key()`?